waves/wallet/src/wallet/decompose_transaction.rs

```diff
@@ -0,0 +1,77 @@
+use crate::{
+    wallet::{current, get_txouts, Wallet},
+    SECP,
+};
+use anyhow::{Context, Result};
+use elements_fun::{AssetId, Transaction, TxOut, UnblindError};
+use futures::lock::Mutex;
+use serde::Serialize;
+use wasm_bindgen::JsValue;
+
+#[derive(Serialize)]
+pub struct TransactionElements {
+    our_inputs: Vec<(AssetId, u64)>,
+    our_outputs: Vec<(AssetId, u64)>,
+}
+
+pub async fn decompose_transaction(
+    name: String,
+    current_wallet: &Mutex<Option<Wallet>>,
+    transaction: Transaction,
+) -> Result<TransactionElements, JsValue> {
+    let wallet = current(&name, current_wallet).await?;
+    let blinding_key = wallet.blinding_key();
+
+    let our_inputs = get_txouts(&wallet, |utxo, ref txout| {
+        transaction
+            .input
+            .iter()
+            .find_map(|txin| {
```

Comment:
I assume you want to `filter_map` here?

Reply:
I don't understand how this is actually giving you a `Vec` :man_shrugging:

Reply:
You wrote `get_txouts`!! :stuck_out_tongue: The way I understand it is that every wallet txout will be filtered by checking if it is part of Bob's transaction, which is why if we find it in Bob's transaction we "add" it to the inputs that will be returned by `get_txouts`.

Reply:
I am confused. From my understanding, `find_map` will only give you the first one and not a list which is why I don't understand how `our_inputs` is a `Vec` again in the end. Is there some magic going on in `collect`?

I think you want to be using `filter_map`, otherwise you will always get a max one input.